storage/src/cloud.ts

```diff
@@ -0,0 +1,139 @@
+import stream from 'stream';
+import { S3 } from 'aws-sdk';
+import { v4 as uuid4 } from 'uuid';
+import {
+  Storage,
+  PutOptions,
+  FileInfo,
+  FileInfoWithContent,
+  UploadToken,
+  InvalidTokenError,
+} from './interfaces';
+
+interface Config {
+  region: string;
+  cdnBaseUrl: string;
+  bucket: string;
+  keyPrefix: string;
+  uploadExpirationMs: number;
+  credentials: {
+    accessKeyId: string,
+    secretAccessKey: string,
+    sessionToken?: string,
+  };
+}
+
+export class CloudStorage implements Storage {
+  constructor(
+    protected readonly config: Config,
+    protected readonly s3: S3 = new S3({
+      region: config.region,
+      credentials: config.credentials,
+    }),
+  ) {
+  }
+
+  public publicUrl(objectId: string): string {
+    return `${this.config.cdnBaseUrl}/${this.config.keyPrefix}/${objectId}`;
+  }
+
+  protected key(objectId: string, prefix: string = ''): string {
+    return `${prefix}${this.config.keyPrefix}/${objectId}`;
+  }
+
+  public async createUpload(opts?: Partial<PutOptions>): Promise<{ token: string, signedUrl: string }> {
+    const objectId = uuid4();
+    const signedUrl = await this.s3.getSignedUrlPromise('putObject', {
+      Bucket: this.config.bucket,
+      Key: this.key(objectId, 'tmp/'),
+      Expires: Math.floor(this.config.uploadExpirationMs / 1000),
```

Comment:
maybe ceil so you won't expire before expected? (expiring after is usually better accepted)

Reply:
Not critical IMHO.